### PR TITLE
[Gekidou - MM-44258] Search Screen - Results Empty State

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR adds the Results screen emtpy state component with options for Empty Files and Empty Messages.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39758

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Screenshots
|  No Files   |  No Messages  |
|----|----|
| image | image |
| image | image |

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: app/screens/home/search/results/empty_files.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            />
+            <Path
+                d='M33.7979 113.687C35.3653 114.135 36.5259 113.978 37.2797 113.215L61.9397 86.1514C62.6935 85.3885 63.0045 84.4077 62.8849 83.1968C62.7652 81.9859 62.0114 80.6296 60.6235 79.1039C59.1159 77.7114 57.7758 76.9122 56.5793 76.7184C55.3828 76.5247 54.4136 76.8758 53.6598 77.7719L26.7385 102.729C25.9847 103.492 25.8651 104.63 26.3676 106.156C26.8701 107.681 27.8154 109.207 29.1913 110.733C30.687 112.259 32.2305 113.239 33.7979 113.687Z'
+                fill='#FFBC1F'
+            />
+            <Path
+                d='M31.9929 97.8425L42.0738 108.007L54.2274 94.6662L45.3132 85.5606L31.9929 97.8425Z'
+                fill='#7A5600'
+            />
+            {/* <defs> */}
```
Comment (jfrerich) [inline app/screens/home/search/results/empty_files.tsx:56]: I am not sure what this portion of the SVG references in the image.  I didn't see any difference when it was added or omitted.

**Comment by matthewbirtch** [comment]
> @matthewbirtch, let me know if you want to add any theming to the SVGs. I did notice the very light shadow underneath the magnifying glass is hard to see in dark mode.

@jfrerich this specific illustration doesn't need any theme colors. The shadow is what it is - won't be so visible on dark themes, but that's ok.

File: app/screens/home/search/results/empty_files.tsx (hunk 1 above)
Comment (matthewbirtch) [reply app/screens/home/search/results/empty_files.tsx:56]: Yep, I think this can be omitted

**Comment by matthewbirtch** [comment]
@jfrerich how do I test this out. Based on the screenshots, it looks like the content of the empty state is sitting too high vertically, but I just checked out the branch and don't see the empty state when trying to search right now.

**Comment by jfrerich** [comment]
I don't have an easy way to test this out right now, but I will add something.  I think the easiest will be to connect clicking the search icon so that it opens the results immediately.  Give me a bit and I'll come back with a new commit

File: app/screens/home/search/results/header.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import {useTheme} from '@context/theme';
+import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
+
+type Props = {
+    onToggle: (val: boolean) => void;
+    showMessages: boolean;
+    numberFiles: number;
+    numberMessages: number;
+}
+const Header = ({onToggle, showMessages, numberFiles, numberMessages}: Props) => {
```
Comment (avinashlng1080) [inline app/screens/home/search/results/header.tsx:17]: I think it is easier to keep track of the tab press and `showMessages`.   The selected `tab` press is not turning blue.  I made some minor refactoring in this file. 
[refactoring.txt](https://github.com/mattermost/mattermost-mobile/files/8706004/refactoring.txt)
.

File: app/screens/home/search/results/header.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            <Button
+                containerStyle={[styles.button, selected ? styles.selectedButton : undefined]}
+                onPress={onPress}
+            >
+                <Text style={[styles.text, showMessages ? styles.selectedText : undefined]}>
+                    {text}
+                </Text >
+            </Button>
+        );
+    };
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply app/screens/home/search/results/header.tsx]: done.

File: app/screens/home/search/results/header.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                    text={`${filesText} (${numberFiles})`}
+                />
+            </View>
+            <View style={styles.divider}/>
+        </>
+
+    );
+};
+
+const getStyleFromTheme = makeStyleSheetFromTheme((theme) => {
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply app/screens/home/search/results/header.tsx:18]: done.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@jfrerich aside from the centering problem, just a couple other things:

For the files empty state, it looks like the wrong text was provided for you. Sorry about that - I guess I missed that in my review a while back. It should be as follows:

> **No files matching “Blah”**
> Check the spelling or try another search

2. The tab styling for `Messages` and `Files` at the top isn't quite right and there is extra space being applied above them. Should I assume these will be fixed in a separate PR? If not, I can provide more specific details about what is off.

**Comment by jfrerich** [comment]
@matthewbirtch, please take another look to verify text and spacing reduced above the tabs buttons.  

We might as well fix the styling in this PR. Thanks.

Something is off when trying to center the empty results vertically.  @elias, maybe you could help me with that.

**Comment by enahum** [comment]
@jfrerich you should look into `filtered_list` that shows the `<NoResultWithTerm>` and also handles the keyboard being displayed or not.

**Comment by jfrerich** [comment]
Thanks @enahum.  I'll look into that component.

**Comment by matthewbirtch** [comment]
@jfrerich I'm not seeing a difference with the tab styling - it should match the tabs used in the threads view. I did notice some mistakes in the Figma file from when Marina was working on this though and that may have caused some confusion. Can you tighten up the tabs? 

image

Design:

**Comment by larkox** [comment]
@matthewbirtch I fixed the styles. Also, now it should react to changes to the search value (you will start seeing a "not implemented" text since that is where the "landing page" of the search will be, which will be implemented on subsequent PRs).

There is something that I am not sure if has been discussed already. The search bar used is the same used in other places, so instead of the cancel, we have a back arrow. Also, the cross is slightly different from the figma designs. Just wanted to make sure before doing any changes to those whether the figma designs are outdated on that aspect, or we really want different looks on the search bar for different places.

File: app/screens/home/search/results/empty_messages.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import React from 'react';
+import Svg, {Path, Ellipse} from 'react-native-svg';
+
+function EmptyMessages() {
```
Comment (enahum) [inline app/screens/home/search/results/empty_messages.tsx:6]: does this component differ from the one we already have under no result with term?

File: app/screens/home/search/results/empty_state.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                values={{
+                    searchValue,
+                }}
+            />
+            <FormattedText
+                defaultMessage={'Check the spelling or try another search.'}
+                id={'screen.search.empty.paragraph'}
+                style={styles.paragraph}
+                testID='search.empty.paragraph'
+            />
```
Comment (enahum) [inline app/screens/home/search/results/empty_state.tsx:69]: this looks to me exactly the same as `NoResultWithTerm` only that the illustration may change

**Review by enahum — CHANGES_REQUESTED** [review]

**Comment by matthewbirtch** [comment]
> There is something that I am not sure if has been discussed already. The search bar used is the same used in other places, so instead of the cancel, we have a back arrow. Also, the cross is slightly different from the figma designs. Just wanted to make sure before doing any changes to those whether the figma designs are outdated on that aspect, or we really want different looks on the search bar for different places.

The back arrow is for android, the 'cancel' button is for iOS

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Styles/spacing for the tabs style needs to be updated:

1. In the expanded state (Default search view), the padding at the top of the tabs needs to be reduced by 9. Green bar in the screenshot here is the correct padding.
image

3. In the collapsed state, the padding at the top of the tabs needs to be increased by 6. The green bar here is the correct padding.

image

3. The Unselected tab is using the wrong color for the label. It should be using 56% center channel color, but it looks like it's using pure black.

4. We need to add `“` and `”` quote marks around the search term in the empty state. (e.g.  No matches for “Blah”

5. The empty state content is now centered when the keyboard is not showing, but when the keyboard _is_ showing, the content is cut off by the keyboard.
IMG_9344
.

File: app/screens/home/search/results/empty_state.tsx (hunk 6 above)
Comment (jfrerich) [reply app/screens/home/search/results/empty_state.tsx:69]: Yes, they appear almost identical.  The ony differentce is the `SearchIllustration` vs `EmptyMessages` and `EmtpyFiles` components.  

How about passing in the `SearchIllustration` as a prop to `NoResultsWithTerm`

**Comment by enahum** [comment]
@matthewbirtch what are the differences between this empty state illustration for posts and the empty state for find channels?

To me they look identical, aren't they?

**Comment by jfrerich** [comment]
@enahum, @matthewbirtch, the wording is slightly differenct for empty files and messages from the wording in emtpy state for find channels.  

Do we want to include the varying titles or use a generic sentence for all?

files: `No files matching "Blah"` 
messages: `No matches found for "Blah"`
default: `No results for Blah`

**Comment by matthewbirtch** [comment]
> @matthewbirtch what are the differences between this empty state illustration for posts and the empty state for find channels?

@enahum the illustrations are the exact same. The file search empty state illustration is different though

**Comment by matthewbirtch** [comment]
> Do we want to include the varying titles or use a generic sentence for all?
> 
> files: `No files matching "Blah"` messages: `No matches found for "Blah"` default: `No results for Blah`

I'd still like to see the files title slightly different, but I we can make the default and messages results title the same. 

files: No files matching “Blah”
messages: No results for “Blah”
default: No results for “Blah”

Please use proper `“` and `”` quote characters - not the `"`

**Comment by enahum** [comment]
> @enahum, @matthewbirtch, the wording is slightly differenct for empty files and messages from the wording in emtpy state for find channels.
> 
> Do we want to include the varying titles or use a generic sentence for all?
> 
> files: `No files matching "Blah"` messages: `No matches found for "Blah"` default: `No results for Blah`

@jfrerich fair enough but I don't like this approach.

This is what I'm seeing: The emptyMessage illustration is exactly the same as the `SearchIllustration` so first of all the was no need for the `EmptyMessage` illustration.

Now, instead of passing the illustration as a prop, I would do something totally different.
in the `<NoResultsWithTerm>` component we could set a prop called `type?: 'default' | 'messages' |  'files'` that will fallback to `default`

if the type is either `default` or `messages` it will use the `<SearchIllustration>` component, in case of `files` then use `<SearchFilesIllustration>` component.

Then based on the `type` prop you can render the appropriate text message.

**Comment by enahum** [comment]
Now that said, I would definitely combine the `messages` and `default`  values and have it render the same message as well.

**Comment by enahum** [comment]
> Please use proper opening “ and closing ” quote characters - not the "

@matthewbirtch could that cause problems for translators ?

**Comment by enahum** [comment]
> > Please use proper opening “ and closing ” quote characters - not the "
> 
> @matthewbirtch could that cause problems for translators ?

@jfrerich perhaps to avoid problems with translators you change it like

```
const NoResultsWithTerm = ({term}: Props) => {
    const theme = useTheme();
    const style = getStyleFromTheme(theme);

    return (
        <View style={style.container}>
            <SearchIllustration/>
            <View style={{flexDirection: 'row'}}>
                <FormattedText
                    id='mobile.no_results_with_term'
                    defaultMessage='No results for'
                    style={style.result}
                />
                <Text style={style.result}>{` “${term}”`}</Text>
            </View>
            <FormattedText
                id='mobile.no_results.spelling'
                defaultMessage='Check the spelling or try another search.'
                style={style.spelling}
            />
        </View>
    );
};
```

**Comment by jfrerich** [comment]
@matthewbirtch, I am not seeing those padding issues in the latest commit.  I did you should see the unselected text color fixed though

**Comment by enahum** [comment]
FYI the keyboard issue is only on iOS, I did share in another comment a reference to a component that deals with this by using the `useKeyboardHeight` hook

**Comment by jfrerich** [comment]
@matthewbirtch, are the padding issues you are seeing also only on ios? I reduced the top margin by 7 for android and we can see the button lines up with the green bar padding. I temporarily made the selected button darker so you can see the contrasting colors easier. 
image

**Comment by enahum** [comment]
Do add me as a reviewer once all the UX concerns have been addressed ;)

**Comment by jfrerich** [comment]
@enahum, were the changes in your commit addressing Matt's concerns?

**Comment by enahum** [comment]
@jfrerich my changes were to make the inner header remain in the position that is set initially, there are many things in this PR that needs addressing.

This fixes were applied while I was in a call with @larkox and I was under the impression that he was going to finish it on Monday.

That said, my fixes also need a fix for Android, which @larkox is already aware of it.

**Comment by larkox** [comment]
Before getting @matthewbirtch into the loop, I would prefer some early feedback from @enahum .

There is a bunch of changes in the code, and some maybe controversial decisions, and changing those may screw the current design, so would prefer a more "stable" state in the code before doing the UX review.

**Comment by enahum** [comment]
@larkox @jfrerich let's jump on a call so you can explain the changes made here and the reason behind them

**Comment by larkox** [comment]
I reverted most of the hook changes, and seems to be working fine. The main caveat is that the search results header is again inside the scroll, so it will not stay always on top. Hopefully, when the changes on the navigation header are done, that can be fixed.

**Comment by larkox** [comment]
I reverted most of the hook changes, and seems to be working fine. The main caveat is that the search results header is again inside the scroll, so it will not stay always on top. Hopefully, when the changes on the navigation header are done, that can be fixed.

**Review by enahum — APPROVED** [review]

**Comment by matthewbirtch** [comment]
This got merged? Sorry, I may have missed part of the conversation on this, but last I looked it wasn't ready to merge.

**Comment by enahum** [comment]
@matthewbirtch at this stage funtionality is more important, pixel perfect will have to be done at a later stage.

We chatted about this in our meeting today

**Pull request merged**